src/markdown.rs

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+// This file is part of Static Web Server.
+// See https://static-web-server.net/ for more information
+// Copyright (C) 2019-present Jose Quintana <joseluisq.net>
+
+//! Markdown content negotiation module.
+//!
+//! This module handles content negotiation for markdown files when the client sends
+//! an Accept header that includes text/markdown.
+
+use headers::HeaderMapExt;
+use hyper::{Body, Request, Response};
+use std::path::Path;
+
+use crate::{
+    Error, fs::meta::try_markdown_variant, handler::RequestHandlerOpts, headers_ext::Accept,
+};
+
+/// Pre-process a request to check if a markdown variant URI should be used.
+/// Returns the modified URI path if a markdown variant exists, None otherwise.
+pub(crate) fn pre_process<T>(
+    _opts: &RequestHandlerOpts,
+    req: &Request<T>,
+    base_path: &Path,
+    uri_path: &str,
+) -> Option<String> {
+    // Check if the client accepts markdown
+    let accepts_markdown = req
+        .headers()
+        .typed_get::<Accept>()
+        .map(|accept| accept.accepts_markdown())
+        .unwrap_or(false);
+
+    if !accepts_markdown {
+        return None;
+    }
+
+    // Construct the full file path
+    let mut file_path = base_path.to_path_buf();
+    let sanitized_path = uri_path.trim_start_matches('/');
+    if !sanitized_path.is_empty() {
+        file_path.push(sanitized_path);
+    }
+
+    // Try to find a markdown variant
+    let md_path = try_markdown_variant(&file_path)?;
+
+    tracing::info!("markdown: found variant {:?}", md_path);
+
+    // Convert the markdown file path back to a URI path
+    // Strip the base_path and prepend '/'
+    md_path
+        .strip_prefix(base_path)
+        .ok()
+        .and_then(|p| p.to_str())
+        .map(|s| format!("/{}", s))
+}
+
+/// Post-process the response to set the correct Content-Type for markdown files.
+pub(crate) fn post_process<T>(
+    opts: &RequestHandlerOpts,
+    req: &Request<T>,
+    mut resp: Response<Body>,
+) -> Result<Response<Body>, Error> {
+    // Only process if markdown negotiation is enabled
+    if !opts.accept_markdown {
+        return Ok(resp);
+    }
+
+    // Check if the client accepts markdown
+    let accepts_markdown = req
+        .headers()
+        .typed_get::<Accept>()
+        .map(|accept| accept.accepts_markdown())
+        .unwrap_or(false);
+
+    if !accepts_markdown {
+        return Ok(resp);
+    }
+
```

Comment:
(1) If I make a `GET` request to fetch an `.html` file, then the markdown headers below are applied.
I suggest refactoring the code to do this below, making sure to apply those headers only to the markdown variant if applicable.

```suggestion
pub(crate) fn post_process(
    is_markdown_variant: bool,
    opts: &RequestHandlerOpts,
    mut resp: Response<Body>,
) -> Result<Response<Body>, Error> {
    if !is_markdown_variant || !opts.accept_markdown {
        return Ok(resp);
    }

```